PolyEngine/Engine/Res/Shaders/blinn-phongFrag.shader

```diff
@@ -44,73 +52,108 @@ in vec3 vNormal;
 
 layout(location = 0) out vec4 color;
 
-vec3 ambientlLight(DiffuseLight ambientLight) {
-	return vec3(ambientLight.Color.rgb * max(ambientLight.Intensity, 0.0));
+vec3 ambientLighting()
+{
+	return uMaterial.Ambient.rgb * uAmbientLight.Color.rgb * uAmbientLight.Intensity;
 }
 
-Phong directionalLight(DirectionalLight light, vec3 positionWS, vec3 normalWS) {
-	Phong result;
-	DiffuseLight base = light.Base;
-	vec3 dir = light.Direction.xyz;
-	float dirNdotL = max(dot(normalWS, dir), 0.0);
-	result.Diffuse = vec3(base.Color.rgb) * max(base.Intensity, 0.0) * dirNdotL;
-	
-	vec3 reflectDir = reflect(-dir, normalWS);
-	float spec = pow(max(dot(uCameraDirection.xyz, reflectDir), 0.0), 32);
-	result.Specular = uSpecularStrength * spec * base.Color.rgb;
-	
-	return result;	
+// returns intensity of diffuse reflection
+vec3 diffuseLighting(in vec3 N, in vec3 L, in vec3 LightColor)
+{
+	// calculation as for Lambertian reflection
+	float NdotL = clamp(dot(N, L), 0.0, 1.0);
+	return vec3(NdotL) * uMaterial.Diffuse.rgb * LightColor;
 }
 
-float pointLightAttenuation(PointLight light, vec3 vertexPos) {
-	float distanceToLight = length(light.Position.xyz - vertexPos);
-	float attentuation = 1.0 / (1.0 + light.Attenuation * pow(distanceToLight, 2));
-	return mix(0.0, attentuation, step(distanceToLight, light.Range));
+// returns intensity of specular reflection
+vec3 specularLighting(in vec3 N, in vec3 L, in vec3 V, in vec3 LightColor)
+{
+	float specularTerm = 0;
+
+	// calculate specular reflection only if
+	// the surface is oriented to the light source
+	if (dot(N, L) > 0)
```

Comment:
We can get rid of branching here by removing if and changing assigment to `specularTerm` like that:
````
const float dotNL = dot(N, L);
specularTerm = ceil(dotNL * dotNL) * pow(dot(N, H), uMaterial.Shininess);

Reply:
We can do that because, assuming N and L are normal vectors, `ceil(dotNL * dotNL)` will always yield result 0  if `dotNL  == 0` and 1 if `dotNL  != 0`.

Reply:
Good idea! Since we care only about positive range of `dot(N, L)` it can be just `ceil(dot(N, L))`:
`specularTerm = ceil(dot(N,L)) * pow(dot(N, H), uMaterial.Shininess);`

checked it with: https://www.shadertoy.com/view/MdjGRy
in GLSL `ceil(x)` on `[-1, 0]` is 0
and specular looks the same on render